python_modules/dagster/dagster/_core/definitions/data_time.py

```diff
@@ -502,6 +502,25 @@ def get_current_data_time(
 
         return min(cast(AbstractSet[datetime.datetime], data_times), default=None)
 
+    def _get_source_data_time(
+        self, asset_key: AssetKey, current_time: datetime.datetime
+    ) -> Optional[datetime.datetime]:
+        latest_record = self.instance_queryer.get_latest_materialization_or_observation_record(
+            AssetKeyPartitionKey(asset_key)
+        )
+        if latest_record is None:
+            return None
+        if latest_record.asset_observation is None:
+            raise DagsterInvariantViolationError(
+                "Cannot calculate data time for source asset with materialization"
+            )
+        # handle case where data_time isn't set
+        return datetime.datetime.utcfromtimestamp(
+            latest_record.event_log_entry.dagster_event.event_specific_data.asset_observation.metadata.get(
+                "data_time"
```

Comment:
I think we'd want to use "dagster/data_time" to indicate that it's a special system piece of metadata.

Reply:
Do we prefix other system metadata?

Reply:
I think this might be the first system metadata? Not confident that we should prefix btw.